### PR TITLE
Constraint-Satisfaction Problems with MiniZinc + Gecode

### LINKED ISSUE #1151: Constraint-Satisfaction Problems with MiniZinc+Geocode (tutorial)
### Brief Summary: 
Constraint-Satisfaction Problems (CSPs) represent problems in terms of variables, their possible values, and constraints on those values. Many problems, from graph-theory to type inference to puzzle-solving, are easily converted into CSPs. In this tutorial, we will use MiniZinc, a programming language built for CSPs, and Geocode, a CSP solver, to model and solve a simple graph-coloring problem.

### Key Takeaways: 
- Understand what a Constraint-Satisfaction Problem is
- Identify problems that lend themselves well to CSP solutions
- Model a CSP in MiniZinc
- Use Geocode to solve the MiniZinc CSP

### References: 
- https://www.cs.upc.edu/~erodri/webpage/cps/theory/cp/intro/slides.pdf
- https://www.minizinc.org/
- https://www.minizinc.org/doc-2.5.3/en/index.html
- https://cis.temple.edu/~giorgio/cis587/readings/constraints.html

### PR DESCRIPTION
**Pull request by nicholaskross**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

Tutorial on using MiniZinc and built-in Gecode to build and solve constraint-satisfaction problems.

This fixes #1151.

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

File: articles/constraint-satisfaction-minizinc-gecode/index.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+author: nicholas-kross
+date: 2021-01-04T00:00:00-20:00
+topics: []
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/constraint-satisfaction-minizinc-gecode/hero.jpg
+    alt: united states map
+---
+Constraint-Satisfaction Problems (CSPs) represent problems as variables, their possible values, and constraints on those values. Many problems, from graph theory to type inference to puzzle-solving, are converted into CSPs. In this tutorial, we will use MiniZinc, a programming language built for CSPs, and Gecode, a CSP solver, to model and solve a simple graph-coloring problem.
```
Comment (botahamec) [inline articles/constraint-satisfaction-minizinc-gecode/index.md:17]: Many of the sentences in the article can be split into shorter sentences for easier reading.

**Review by botahamec — COMMENTED** [review]
@nicholaskross Good job on your first article. I only have a few comments. Once you fix those you should be good to go.

File: articles/constraint-satisfaction-minizinc-gecode/index.md (hunk 1 above)
Comment (nicholaskross) [reply articles/constraint-satisfaction-minizinc-gecode/index.md]: Alright I think I fixed those!

File: articles/constraint-satisfaction-minizinc-gecode/index.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+A domain is, itself, a set of (a finite number of) possible values for a corresponding variable. So if *X={x<sub>1</sub>, x<sub>2</sub>, x<sub>2</sub>}* is your set of variables, then *D*={*d<sub>1</sub>*, *d<sub>2</sub>*, *d<sub>3</sub>*} is your set of domains, where the *i*th variable *x<sub>i</sub>* can be a value from the set *d<sub>i</sub>*.
+
+The final set, *C*, is the set of **constraints** on the values of the variables. A constraint is some rule placed on the value of one or more variables. So if *C*={*x<sub>1</sub>* + *x<sub>2</sub>* = 3}, we can only pick values from *d<sub>1</sub>* and *d<sub>2</sub>* that make *x<sub>1</sub>* and *x<sub>2</sub>* add up to 3.
+
+Putting this all together, we could define a simple CSP as follows:
+
+*X*={*x*, *y*}
+
+All domains are the same: *d<sub>x</sub>*=*d<sub>y</sub>*={2,3}
```
[the remark this answers is not in the document]
Comment (nicholaskross) [reply articles/constraint-satisfaction-minizinc-gecode/index.md]: I think my revision explains it better. (If I'm wrong, let me know).

**Comment by nicholaskross** [comment]
> @nicholaskross Good job on your first article. I only have a few comments. Once you fix those you should be good to go.

Alright, I've committed revisions!

**Review by botahamec — APPROVED** [review]
@hectorkambow This article is ready

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
👋  @nicholaskross  Just wanted to make sure the .mzn file isn't needed in the PR? 

cc @botahamec

**Comment by botahamec** [comment]
> 👋  @nicholaskross  Just wanted to make sure the .mzn file isn't needed in the PR? 
> 
> cc @botahamec 
> 

The mzn file is meant to be a downloadable file for the reader to use.

**Comment by hectorkambow** [comment]
> > 👋  @nicholaskross  Just wanted to make sure the .mzn file isn't needed in the PR?
> > cc @botahamec
> 
> The mzn file is meant to be a downloadable file for the reader to use.

🤔  

Can we somehow place that in a repo and reference as a link? I think that would be the best practice. 

@nicholaskross

**Comment by nicholaskross** [comment]
> > > 👋  @nicholaskross  Just wanted to make sure the .mzn file isn't needed in the PR?
> > > cc @botahamec
> > 
> > 
> > The mzn file is meant to be a downloadable file for the reader to use.
> 
> 🤔
> 
> Can we somehow place that in a repo and reference as a link? I think that would be the best practice.
> 
> @nicholaskross

@hectorkambow I could do that! Here is my new repository: https://github.com/nicholaskross/constraint-satisfaction-minizinc-example

Now I'll make commits with the links.

**Pull request merged**